architecture/2024-10-deploymenttemplate-controller.md

```diff
@@ -0,0 +1,421 @@
+kubectl apply -f app.yaml
+```
+
+#### `app.yaml`
+```yaml
+kind: DeploymentTemplate
+apiVersion: radapp.io/v1alpha3
+metadata:
+  name: app.bicep
+  namespace: radius-system
+spec:
+  template: |
+    {
+      "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+      "languageVersion": "2.1-experimental",
+      "contentVersion": "1.0.0.0",
+      "metadata": {
+        "_EXPERIMENTAL_WARNING": "This template uses ARM features that are experimental. Experimental features should be enabled for testing purposes only, as there are no guarantees about the quality or stability of these features. Do not enable these settings for any production usage, or your production environment may be subject to breaking.",
+        "_EXPERIMENTAL_FEATURES_ENABLED": [
+          "Extensibility"
+        ],
+        "_generator": {
+          "name": "bicep",
+          "version": "0.29.47.4906",
+          "templateHash": "14156924711842038952"
+        }
+      },
+      "parameters": {
+        "tag": {
+          "type": "string"
+        },
+        "port": {
+          "type": "int"
+        }
+      },
+      "imports": {
+        "Radius": {
+          "provider": "Radius",
+          "version": "latest"
+        }
+      },
+      "resources": {
+        "container": {
+          "import": "Radius",
+          "type": "Applications.Core/containers@2023-10-01-preview",
+          "properties": {
+            "name": "container",
+            "properties": {
+              "application": "[parameters('application')]",
+              "container": {
+                "image": "[format('ghcr.io/radius-project/samples/demo:{0}', parameters('tag'))]",
+                "ports": {
+                  "web": {
+                    "containerPort": "[parameters('port')]"
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  parameters: |
+    {
+      "tag": {
+        "value": "latest"
+      },
+      "port": {
+        "value": 3000
+      }
+    }
+```
+
+## Design
+
+### High Level Design
+
+This design proposes the creation of two new Kubernetes custom resources and their corresponding reconcilers: `DeploymentTemplate` and `DeploymentResource`. The `DeploymentTemplate` controller will be responsible for deploying and reconciling resources defined in ARM JSON manifests via Radius, while the `DeploymentResource` controller will be responsible for reconciling individual resources. The `DeploymentTemplate` controller will have a control loop that will check the status of an in-progress operation, process deletion, and process creation or update. The `DeploymentResource` controller will have a similar control loop that will check the status of an in-progress operation or process deletion if necessary.
+
+### Architecture Diagram
+
+![Architecture Diagram](2024-10-deploymenttemplate-controller/architecture.png)
```

Comment:
At this stage, is the deployment template in the form of ARM json rather than bicep?
![image](https://github.com/user-attachments/assets/a9ff8c7e-0607-431b-bc83-baeaab42c377)

Reply:
yes, Bicep compilation is done by the CLI in this design doc

Reply:
I'll clear this up on the diagram